Packs/SMIME_Messaging/Integrations/SMIME_Messaging/SMIME_Messaging.py

```diff
@@ -173,112 +330,226 @@ def decode_str(decrypted_text: bytes, encoding: str) -> Tuple[str, str]:
+
     msg = ''
     client.smime.load_key(client.private_key_file, client.public_key_file)
     try:
-        p7, data = SMIME.smime_load_pkcs7(encrypt_message['path'])
+        p7 = SMIME.smime_load_pkcs7(encrypt_message['path'])
+        if isinstance(p7, tuple):
+            p7 = p7[0]
         decrypted_text = client.smime.decrypt(p7)
+        if not decrypted_text:
+            raise ValueError('Unknown error: failed to decrypt message')
         out, msg = decode_str(decrypted_text, encoding)
 
     except SMIME.SMIME_Error as e:
 
         if str(e) == 'no content type':  # If no content type; see if we can process as DER format
-            with open(encrypt_message['path'], "rb") as message_file:
+            demisto.debug('No content type found in message, testing if it is in DER format (binary)')
+            with open(encrypt_message['path'], 'rb') as message_file:
                 p7data = message_file.read()
             p7bio = BIO.MemoryBuffer(p7data)
-            p7 = SMIME.PKCS7(m2.pkcs7_read_bio_der(p7bio._ptr()))
+            p7 = SMIME.load_pkcs7_bio_der(p7bio)
             decrypted_text = client.smime.decrypt(p7, flags=SMIME.PKCS7_NOVERIFY)
+            if not decrypted_text:
+                raise ValueError('Unknown error: failed to decrypt message')
             out, msg = decode_str(decrypted_text, encoding)
 
         else:
             raise
 
-    entry_context = {
-        'SMIME.Decrypted': {
-            'Message': out
-        }
-    }
+    msg_out = out
     human_readable = f'{msg}The decrypted message is: \n{out}'
-
-    return human_readable, entry_context
-
-
-def sign_and_encrypt(client: Client, args: Dict):
-
-    message = args.get('message', '').encode('utf-8')
-    msg_bio = BIO.MemoryBuffer(message)
-    sign = client.private_key_file
-    encrypt = client.public_key_file
+    html_readable = ''
+    if not raw_output:  # Return message after parsing html/images/attachments
+        demisto.debug(f'parsing message:\n\n{out}')
+        [email_body, email_html] = parse_multipart_message(out)
+        if email_html:
+            msg_out = email_html
+            human_readable = f'{msg}Message decrypted successfully'
+            html_readable = email_html
+        elif email_body:
+            msg_out = email_body
+            human_readable = f'### {msg}The decrypted message is: \n ___ \n {email_body}'
+        else:
+            human_readable = f'### {msg}The decrypted message is signed, verify to get the original message\n ___ \n{out}'
+
+    results = [CommandResults(
+        readable_output=human_readable,
+        outputs_prefix='SMIME.Decrypted',
+        outputs={'Message': msg_out},
+        tags=tags,
+    )]
+
+    if html_readable:
+        results.append(CommandResults(
+            raw_response=html_readable,
+            content_format=EntryFormat.HTML,
+            entry_type=EntryType.NOTE,
+        ))
+
+    return results
+
+
+def sign_and_encrypt(client: Client, args: dict):
```

Comment:
The function is very long and I think breaking it down into functions will enhance readability and maintenance.